File: react_front_end/src/services/Firestore.js
import dotenv from "dotenv";
import * as admin from "firebase-admin";

dotenv.config();

export default class Firestore {
  static firebase;
  static firestore;

  static getInstance = () => {
    if (admin.apps.length < 1) {
      const { firebase, firestore } = this.initialise();
      this.firebase = firebase;
      this.firestore = firestore;
    }

    this.firebase = admin;
    this.firestore = admin.app("VIR-US").firestore();

    return this;
  };

  static initialise = () => {
    // Web app's Firebase configuration
    var credentials = {
      type: process.env.REACT_APP_type,
      project_id: process.env.REACT_APP_project_id,
      private_key_id: process.env.REACT_APP_private_key_id,
      private_key: process.env.REACT_APP_private_key.replace(/\\n/g, "\n"),
      client_email: process.env.REACT_APP_client_email,
      client_id: process.env.REACT_APP_client_id,
      auth_uri: process.env.REACT_APP_auth_uri,
      token_uri: process.env.REACT_APP_token_uri,
      auth_provider_x509_cert_url:
        process.env.REACT_APP_auth_provider_x509_cert_url,
      client_x509_cert_url: process.env.REACT_APP_client_x509_cert_url,
    };

    const app = admin.initializeApp(
      {
        credential: admin.credential.cert(credentials),
        databaseURL: "https://stv-game-db-test.firebaseio.com",
      },
      "VIR-US"
    );

    const defaultFirebase = admin;
    const defaultFirestore = admin.firestore(app);

    return {
      firebase: defaultFirebase,
      firestore: defaultFirestore,
    };
  };

  static tribe_from_id = async (game = null, id = null) => {
    return await (
      await this.firestore.doc(`games/${game}/tribes/${id}`).get()
    ).data();
  };

  static count_players = async ({
    game = null,
    from_tribe = null,
    from_team = null,
  }) => {
    if (from_tribe) {
      return (
        await this.firestore.doc(`games/${game}/tribes/${from_tribe}`).get()
      ).data().count_players;
    } else if (from_team) {
      return (
        await this.firestore.doc(`games/${game}/teams/${from_team}`).get()
      ).data().count_players;
    } else {
      return (await this.firestore.doc(`games/${game}`).get()).data()
        .count_players;
    }
  };

  static count_teams = async ({
    game = null,
    from_tribe = null,
    active_team_predicate_value = true,
  }) => {
    let query;
    if (from_tribe) {
      query = this.firestore.doc(`games/${game}/tribes/${from_tribe}`);
    } else {
      query = this.firestore.doc(`games/${game}`);
    }
    return (await query.get()).data().count_teams;
  };

  static batch_update_tribe = async ({
    game = null,
    from_tribe = null,
    to_tribe = null,
  }) => {
    const teams = this.firestore
      .collection(`games/${game}/teams`)
      .where("tribe_id", "==", from_tribe);

    const players = this.firestore
      .collection(`games/${game}/players`)
      .where("tribe_id", "==", from_tribe);

    const player_count = await this.count_players({
      game,
      from_tribe,
    });

    const team_count = await this.count_teams({
      game,
      from_tribe,
    });

    const batch1 = this.firestore.batch();

    (await teams.get()).forEach(async (document_iter) => {
      batch1.update(document_iter.ref, { tribe_id: to_tribe });
    });

    (await players.get()).forEach(async (document_iter) => {
      batch1.update(document_iter.ref, { tribe_id: to_tribe });
    });

    await batch1.commit();

    const batch2 = this.firestore.batch();

    let ref = this.firestore.doc(`games/${game}/tribes/${to_tribe}`);

    let count = (await ref.get()).data().count_players;

    let total = count + player_count;

    batch2.update(ref, { count_players: total });

    ref = this.firestore.doc(`games/${game}/tribes/${to_tribe}`);

    count = (await ref.get()).data().count_teams;

    total = count + team_count;

    batch2.update(ref, { count_teams: total });

    ref = this.firestore.doc(`games/${game}/tribes/${from_tribe}`);

    count = (await ref.get()).data().count_players;

    total = count - player_count;

    batch2.update(ref, { count_teams: total });

    ref = this.firestore.doc(`games/${game}/tribes/${from_tribe}`);

    count = (await ref.get()).data().count_teams;

    total = count - team_count;

    batch2.update(ref, { count_teams: total });

    return await batch2.commit();
  };

  static player_from_id = async ({ game = null, id = null }) => {
    return (
      await this.firestore.doc(`games/${game}/players/${id}`).get()
    ).data();
  };

  static team_from_id = async ({ game = null, id = null }) => {
    return (await this.firestore.doc(`games/${game}/teams/${id}`).get()).data();
  };

  static count_votes = async ({
    game = null,
    from_team = null,
    is_for_win = false,
  }) => {
    let player_counts = {};

    const query = this.firestore.collection(`games/${game}/votes`);

    if (from_team) {
      query.where("team_id", "==", from_team);

      const voteFromIds = [];
      (await query.get()).forEach((vote) => {
        voteFromIds.push(vote.data());
      });

      const playerPromises = voteFromIds.map(async ({ from_id }) => {
        return this.player_from_id({ game, id: from_id });
      });
      const players = await Promise.all(playerPromises);

      const teamPromises = players.map((player) => {
        return this.team_from_id({ game, id: player.team_id });
      });
      const teams = await Promise.all(teamPromises);

      for (let i = 0; i < teams.length; i++) {
        const team = teams[i];
        const player = players[i];
        const vote = voteFromIds[i];

        const value = player_counts[vote.to_id];
        if (isNaN(value)) {
          player_counts[vote.to_id] = 1;
        } else {
          player_counts[vote.to_id] += 1;
        }
      }
    } else {
      const voteFromIds = [];
      (await query.get()).forEach((vote) => {
        const data = vote.data();

        const value = player_counts[data.to_id];
        if (isNaN(value)) {
          player_counts[data.to_id] = 1;
        } else {
          player_counts[data.to_id] += 1;
        }
      });
    }

    return player_counts;
  };

  static get_game_info = async ({ game = null }) => {
    if (!game) return false;
    const response = (
      await this.firestore.collection(`games`).doc(`${game}`).get()
    ).data();
    return response;
  };

  static add_game = async ({ game = null, hashtag = null, testId = null }) => {
    if (!game) return false;
    if (!hashtag) return false;

    const response = await this.firestore.collection(`games`).add({
      game,
      hashtag,
      count_players: 0,
      count_teams: 0,
      count_tribes: 0,
    });

    const map = {
      id: testId ? testId : response.id,
      ...(await response.get()).data(),
    };

    await response.set(map);

    return response.id;
  };

  static add_challenge = async ({
    game = null,
    name = null,
    message = null,
    testId = null,
  }) => {
    if (!game) return false;
    if (!name) return false;
    if (!message) return false;

    const response = await this.firestore
      .collection(`games`)
      .doc(`${game}`)
      .collection(`challenges`)
      .add({
        name,
        message,
        start_timestamp: Date.now(),
        end_timestamp: Date.now() + 10080000,
      });

    const map = {
      id: testId ? testId : response.id,
      ...(await response.get()).data(),
    };

    await response.set(map);

    return (await response.get()).data();
  };

  static add_submission_entry = async ({
    game = null,
    likes = null,
    views = null,
    player_id = null,
    team_id = null,
    tribe_id = null,
    challenge_id = null,
    url = null,
    testId = null,
  }) => {
    if (
      (!game || !likes || !views || !player_id,
      !team_id,
      !tribe_id || !challenge_id || !url)
    )
      return false;

    const response = await this.firestore
      .collection(`games`)
      .doc(`${game}`)
      .collection(`entries`)
      .add({
        likes,
        views,
        player_id,
        team_id,
        tribe_id,
        challenge_id,
        url,
      });

    const map = {
      id: testId ? testId : response.id,
      ...(await response.get()).data(),
    };

    await response.set(map);

    return (await response.get()).data();
  };

  static add_player = async ({
    game = null,
    tiktok = null,
    email = null,
    tribe_id = null,
    team_id = null,
    active = null,
    testId = null,
    phone = null,
    code = null,
  }) => {
    if (
      !game ||
      !tiktok ||
      !email ||
      !tribe_id ||
      !team_id ||
      !active ||
      !phone ||
      !code
    )
      return false;

    const response = await this.firestore
      .collection(`games`)
      .doc(`${game}`)
      .collection(`players`)
      .add({
        tiktok,
        email,
        tribe_id,
        team_id,
        active,
        phone,
        code,
      });

    const map = {
      id: testId ? testId : response.id,
    };

    await response.update(map);

    const data = (await response.get()).data();

    return data;
  };

  static add_vote = async ({
    game = null,
    from_id = null,
    to_id = null,
    team_id = null,
    is_for_win = null,
    testId = null,
  }) => {
    if (!game || !from_id || !to_id || !team_id || !is_for_win) return false;

    const response = await this.firestore
      .collection(`games`)
      .doc(`${game}`)
      .collection(`votes`)
      .add({
        from_id,
        to_id,
        team_id,
        is_for_win,
      });

    const map = {
      id: testId ? testId : response.id,
      ...(await response.get()).data(),
    };

    await response.set(map);

    return (await response.get()).data();
  };

  static verify_code = async ({ phone = null, code = null, game = null }) => {
    if (!phone || !code || !game) return false;

    const player = this.firestore
      .collection(`games/${game}/players`)
      .where("phone", "==", phone)
      .where("code", "==", code)
      .limit(1);

    const query = await player.get();

    if (!query.docs.length) throw new Error("player not found");

    const data = query.docs[0];

    data.ref.update({ active: true });

    return "verified";
  };
}

File: backend/game_engine/firestore.py
from game_engine.database import Database, Data
import attr
from typing import Dict, Iterable, Tuple, Optional
from game_engine.database import Player, Team, Tribe
from game_engine.database import Challenge, Entry, Vote, Game, Ballot
from multiprocessing import Pool
from itertools import product
import firebase_admin
from firebase_admin import credentials, firestore
from google.cloud.firestore_v1 import Client, Increment, Query
from google.cloud.firestore_v1.document import DocumentReference
import logging
import sys
from concurrent.futures import ThreadPoolExecutor
import copy
import json
import uuid
import datetime
import time

# TODO(brandon): change Data interface to use counters instead of size
# by convention.
_THREAD_POOL_SIZE = 100

logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)


class FirestoreData(Data):
    """This class wraps firebase objects with dynamically assigned
    properties in order to avoid having to call object.get(property)
    on every access."""

    def __init__(self, document: DocumentReference):
        setattr(self, 'id', document.id)
        document_dict = document.to_dict()
        if document_dict:
            for k, v in document_dict.items():
                setattr(self, k, v)


class FirestoreGame(FirestoreData, Game):
    pass


class FirestorePlayer(FirestoreData, Player):
    pass


class FirestoreVote(FirestoreData, Vote):
    pass


class FirestoreTeam(FirestoreData, Team):
    pass


class FirestoreTribe(FirestoreData, Tribe):
    pass


class FirestoreChallenge(FirestoreData, Challenge):
    pass


class FirestoreEntry(FirestoreData, Entry):
    pass


class FirestoreBallot(FirestoreData, Ballot):
    pass


class FirestoreDataStream(object):
    def __init__(self, stream: Iterable):
        self._stream = stream

    def __iter__(self):
        return self

    def __next__(self):
        return FirestoreData(self._stream.__next__())


class FirestorePlayerStream(FirestoreDataStream):
    def __next__(self):
        return FirestorePlayer(self._stream.__next__())


class FirestoreVoteStream(FirestoreDataStream):
    def __next__(self):
        return FirestoreVote(self._stream.__next__())


class FirestoreTeamStream(FirestoreDataStream, Team):
    def __next__(self):
        return FirestoreTeam(self._stream.__next__())


class FirestoreChallengeStream(FirestoreDataStream):
    def __next__(self):
        return FirestoreChallenge(self._stream.__next__())


class FirestoreEntryStream(FirestoreDataStream):
    def __next__(self):
        return FirestoreEntry(self._stream.__next__())


class FirestoreDB(Database):

    def __init__(self, json_config_path: str, game_id: str = None):
        cred = credentials.Certificate(json_config_path)
        if not firebase_admin._apps:
            firebase_admin.initialize_app(cred)
        self._game_id = game_id if game_id else self._create_game_id()
        self._client = firestore.client()
        self._thread_pool_size = _THREAD_POOL_SIZE

    def _create_game_id(self):
        return ""

    def import_collections(self, collections_json: str) -> None:
        """Function for restoring test DB data."""

        batch = self._client.batch()
        collections_dict = json.loads(collections_json)
        for path in collections_dict:
            for document_id in collections_dict[path]:
                document_ref = self._client.document(
                    "{}/{}".format(path, document_id))
                properties_dict = collections_dict[path][document_id]
                properties_dict['id'] = document_id
                batch.set(document_ref, properties_dict, merge=True)
        batch.commit()

    def export_collections(self, collection_paths):
        """Function for persisting test DB data."""

        paths = collection_paths
        collections = dict()
        dict4json = dict()
        doc_count = 0

        for path in paths:
            collections[path] = self._client.collection(path).stream()
            dict4json[path] = {}
            for document in collections[path]:
                docdict = document.to_dict()
                dict4json[path][document.id] = docdict
                doc_count += 1

        return json.dumps(dict4json)

    def delete_collection(self, path, batch_size: int = 10):
        coll_ref = self._client.collection(path)
        docs = coll_ref.limit(batch_size).stream()
        deleted = 0

        for doc in docs:
            print(f'Deleting doc {doc.id} => {doc.to_dict()}')
            doc.reference.delete()
            deleted = deleted + 1

        if deleted >= batch_size:
            return self.delete_collection(path, batch_size)

    @classmethod
    def add_game(cls, json_config_path: str, hashtag: str, country_code: str = 'US', max_reschedules: int = 5) -> str:
        cred = credentials.Certificate(json_config_path)
        if not firebase_admin._apps:
            firebase_admin.initialize_app(cred)
        client = firestore.client()
        game_ref = client.collection('games').document()
        game_ref.set({
            'count_players': 0,
            'count_teams': 0,
            'count_tribes': 0,
            'country_code': country_code,
            'game_has_started': False,
            'max_reschedules': max_reschedules,
            'times_rescheduled': 0,
            'last_checked_date': '2020-01-01',
            'game':  hashtag,
            'hashtag': hashtag,
            'id': game_ref.id
        })
        return str(game_ref.id)

    @classmethod
    def add_user(cls, json_config_path: str, name: str, tiktok: str, phone_number: str, game_id: str = None) -> None:
        cred = credentials.Certificate(json_config_path)
        if not firebase_admin._apps:
            firebase_admin.initialize_app(cred)
        client = firestore.client()
        users = client.collection('users').where(
            'phone_number', '==', phone_number).get()
        if len(users) == 0:
            user_ref = client.collection('users').document()
        else:
            user_ref = users[0].reference
        user_ref.set({
            'name': name,
            'tiktok': tiktok,
            'phone_number': phone_number,
            'id': user_ref.id,
            'game_id': game_id
        })
        return user_ref.id

    def add_challenge_entry(self, entry: Entry) -> None:
        entry_ref = self._client.collection('games/{}/entries').document()
        entry_ref.set({
            'likes': entry.likes,
            'views': entry.views,
            'player_id': entry.player_id,
            'tribe_id': entry.tribe_id,
            'challenge_id': entry.challenge_id,
            'team_id': entry.team_id,
            'url': entry.url
        })
        return str(entry_ref.id)

    def add_challenge(self, challenge: Challenge) -> None:
        challenge_ref = self._client.collection(
            f'games/{self._game_id}/challenges').document()
        challenge_ref.set({
            'name': challenge.name,
            'message': challenge.message,
            'start_timestamp': challenge.start_timestamp,
            'end_timestamp': challenge.end_timestamp,
            'complete': challenge.complete,
            'id': str(challenge_ref.id)
        })
        return str(challenge_ref.id)

    def _tribe_update_fn(self, stream: Iterable, updates_dict: Dict, batch_size=500):
        batch = self._client.batch()
        document_iter = iter(stream)
        for _ in range(batch_size):
            try:
                document_ref = next(document_iter).reference
                batch.update(document_ref, updates_dict)
            except StopIteration:
                break
        batch.commit()

    def batch_update_tribe(self, from_tribe: Tribe, to_tribe: Tribe) -> None:
        # TODO(brandon): consider strengthening this because if the server crashes
        # mid-merge the counts will be difficult to fix.
        teams = self._client.collection('games/{}/teams'.format(self._game_id)).where(
            'tribe_id', '==', from_tribe.id).stream()
        players = self._client.collection('games/{}/players'.format(self._game_id)).where(
            'tribe_id', '==', from_tribe.id).stream()
        player_count = self.count_players(from_tribe=from_tribe)
        team_count = self.count_teams(from_tribe=from_tribe)
        with ThreadPoolExecutor(max_workers=self._thread_pool_size) as executor:
            executor.submit(self._tribe_update_fn, teams,
                            {"tribe_id": to_tribe.id})
            executor.submit(self._tribe_update_fn, players,
                            {"tribe_id": to_tribe.id})

        batch = self._client.batch()
        batch.update(self._client.document('games/{}/tribes/{}'.format(self._game_id, to_tribe.id)), {
            'count_players': Increment(player_count)
        })

        batch.update(self._client.document('games/{}/tribes/{}'.format(self._game_id, to_tribe.id)), {
            'count_teams': Increment(team_count)
        })

        batch.update(self._client.document('games/{}/tribes/{}'.format(self._game_id, to_tribe.id)), {
            'size': Increment(player_count)
        })

        batch.update(self._client.document('games/{}/tribes/{}'.format(self._game_id, from_tribe.id)), {
            'count_players': Increment(-1 * player_count)
        })

        batch.update(self._client.document('games/{}/tribes/{}'.format(self._game_id, from_tribe.id)), {
            'count_teams': Increment(-1 * team_count)
        })

        batch.update(self._client.document('games/{}/tribes/{}'.format(self._game_id, from_tribe.id)), {
            'size': Increment(-1 * player_count)
        })

        batch.commit()

    def stream_entries(self, from_tribe: Tribe = None, from_team: Team = None,
                       from_challenge: Challenge = None) -> Iterable[Entry]:
        query = self._client.collection(
            'games/{}/entries'.format(self._game_id))
        if from_challenge:
            query = query.where('challenge_id', '==', from_challenge.id)
        if from_tribe:
            query = query.where('tribe_id', '==', from_tribe.id)
        if from_team:
            query = query.where('team_id', '==', from_team.id)
        return FirestoreEntryStream(query.stream())

    def stream_teams(self, from_tribe: Tribe,
                     team_size_predicate_value: [int, None] = None,
                     order_by_size=True,
                     descending=False
                     ) -> Iterable[Team]:
        query = self._client.collection('games/{}/teams'.format(self._game_id)).where(
            'tribe_id', '==', from_tribe.id)
        if team_size_predicate_value:
            query = query.where('count_players', '==',
                                team_size_predicate_value)
        elif order_by_size:
            query = query.order_by(
                'count_players', direction=Query.DESCENDING if descending else Query.ASCENDING)
        return FirestoreTeamStream(stream=query.stream())

    def stream_players(self, active_player_predicate_value=True) -> Iterable[Player]:
        query = self._client.collection('games/{}/players'.format(self._game_id)).where(
            'active', '==', active_player_predicate_value
        )
        return FirestorePlayerStream(stream=query.stream())

    def count_players(self, from_tribe: Tribe = None, from_team: Team = None) -> int:
        if from_tribe:
            return self._client.document('games/{}/tribes/{}'.format(self._game_id, from_tribe.id)).get().get('count_players')

        elif from_team:
            return self._client.document('games/{}/teams/{}'.format(self._game_id, from_team.id)).get().get('count_players')
        else:
            return self._client.document('games/{}'.format(self._game_id)).get().get('count_players')

    # TODO(brandon): remove active_team_predicate_value argument. unused and makes counters harder
    # at scale.
    def count_teams(self, from_tribe: Tribe = None, active_team_predicate_value=True) -> int:
        if from_tribe:
            query = self._client.document(
                'games/{}/tribes/{}'.format(self._game_id, from_tribe.id))
        else:
            query = self._client.document('games/{}'.format(self._game_id))
        return query.get().get('count_teams')

    def count_votes(self, from_team: Team = None, is_for_win: bool = False) -> Dict[str, int]:
        player_counts = {}

        query = self._client.collection(
            'games/{}/votes'.format(self._game_id))
        if from_team:
            query = query.where('team_id', '==', from_team.id)

            for vote in FirestoreVoteStream(query.stream()):
                voter = self.player_from_id(vote.from_id)
                team = self.team_from_id(voter.team_id)
                if team.id != from_team.id or not voter.active:
                    continue

                if vote.to_id not in player_counts:
                    player_counts[vote.to_id] = 1
                else:
                    player_counts[vote.to_id] = player_counts[vote.to_id] + 1
        else:
            for vote in FirestoreVoteStream(query.stream()):
                if not vote.is_for_win:
                    continue

                if vote.to_id not in player_counts:
                    player_counts[vote.to_id] = 1
                else:
                    player_counts[vote.to_id] = player_counts[vote.to_id] + 1

        return player_counts

    def list_challenges(self, challenge_completed_predicate_value=False) -> Iterable[Challenge]:
        query = self._client.collection(
            'games/{}/challenges'.format(self._game_id)).where(
                'complete', '==', challenge_completed_predicate_value)
        challenge_list = []
        for doc in query.stream():
            fc = FirestoreChallenge(document=doc)
            challenge_list.append(fc)
        return challenge_list

    def list_players(self, from_team: Team, active_player_predicate_value=True) -> Iterable[Player]:
        query = self._client.collection(
            'games/{}/players'.format(self._game_id)).where('team_id', '==', from_team.id).where(
                'active', '==', active_player_predicate_value)
        return FirestorePlayerStream(query.stream())

    def list_teams(self, active_team_predicate_value=True) -> Iterable[Team]:
        query = self._client.collection(
            'games/{}/teams'.format(self._game_id)).where(
                'active', '==', active_team_predicate_value)
        return FirestoreTeamStream(query.stream())

    def game_from_id(self, id: str) -> Game:
        return FirestoreGame(self._client.document("games/{}".format(self._game_id)).get())

    def player_from_id(self, id: str) -> Player:
        return FirestorePlayer(self._client.document("games/{}/players/{}".format(self._game_id, id)).get())

    def team_from_id(self, id: str) -> Team:
        return FirestoreTeam(self._client.document("games/{}/teams/{}".format(self._game_id, id)).get())

    def tribe_from_id(self, id: str) -> Tribe:
        return FirestoreTribe(self._client.document("games/{}/tribes/{}".format(self._game_id, id)).get())

    def challenge_from_id(self, id: str) -> Challenge:
        return FirestoreChallenge(self._client.document("games/{}/challenges/{}".format(self._game_id, id)).get())

    def deactivate_player(self, player: Player) -> None:
        batch = self._client.batch()
        player_ref = self._client.document(
            "games/{}/players/{}".format(self._game_id, player.id))
        tribe_ref = self._client.document(
            "games/{}/tribes/{}".format(self._game_id, player.tribe_id))
        team_ref = self._client.document(
            "games/{}/teams/{}".format(self._game_id, player.team_id))
        game_ref = self._client.document("games/{}".format(self._game_id))
        batch.update(player_ref, {
            'active': False
        })
        batch.update(tribe_ref, {
            'count_players': Increment(-1)
        })
        batch.update(tribe_ref, {
            'size': Increment(-1)
        })
        batch.update(team_ref, {
            'count_players': Increment(-1)
        })
        batch.update(team_ref, {
            'size': Increment(-1)
        })
        batch.update(game_ref, {
            'count_players': Increment(-1)
        })
        users = self._client.collection("users").where(
            "phone_number", "==", player.phone_number).get()
        if len(users) > 0:
            user_ref = users[0].reference
            batch.update(user_ref, {
                'game_id': None
            })
        batch.commit()

    def deactivate_team(self, team: Team) -> None:
        batch = self._client.batch()
        team_ref = self._client.document(
            "games/{}/teams/{}".format(self._game_id, team.id))
        tribe_ref = self._client.document(
            "games/{}/tribes/{}".format(self._game_id, team.tribe_id))
        game_ref = self._client.document("games/{}".format(self._game_id))

        batch.update(team_ref, {
            'active': False
        })

        batch.update(tribe_ref, {
            'count_teams': Increment(-1)
        })

        batch.update(game_ref, {
            'count_teams': Increment(-1)
        })

        batch.commit()

    def save(self, data: Data) -> None:
        properties_dict = copy.deepcopy(data.__dict__)
        if '_document' in properties_dict:
            del properties_dict['_document']
        if isinstance(data, Player):
            self._client.document("games/{}/players/{}".format(self._game_id, data.id)).set(
                properties_dict
            )
        elif isinstance(data, Team):
            self._client.document("games/{}/teams/{}".format(self._game_id, data.id)).set(
                properties_dict
            )
        elif isinstance(data, Tribe):
            self._client.document("games/{}/tribes/{}".format(self._game_id, data.id)).set(
                properties_dict
            )
        elif isinstance(data, Challenge):
            self._client.document("games/{}/challenges/{}".format(self._game_id, data.id)).set(
                properties_dict
            )

    def tribe(self, name: str) -> Tribe:
        tribe_ref = self._client.collection(
            "games/{}/tribes".format(self._game_id)).document()
        tribe_ref.set({
            'name': name,
            'count_players': 0,
            'count_teams': 0,
            'size': 0,
            'active': True,
            'id': tribe_ref.id
        })
        return FirestoreData(tribe_ref.get())

    def player(self, name: str, tiktok: str = None, phone_number: str = None) -> Player:
        batch = self._client.batch()
        player_ref = self._client.collection(
            "games/{}/players".format(self._game_id)).document()
        batch.set(player_ref, {
            'name': name,
            'active': True,
            'tiktok': tiktok,
            'phone_number': phone_number
        })

        game_ref = self._client.document("games/{}".format(self._game_id))
        batch.update(game_ref, {
            'count_players': Increment(1)
        })

        batch.commit()
        return FirestoreData(player_ref.get())

    def _delete_vote_fn(self, stream: Iterable, batch_size=500):
        batch = self._client.batch()
        document_iter = iter(stream)
        for _ in range(batch_size):
            try:
                document_ref = next(document_iter).reference
                batch.delete(document_ref)
            except StopIteration:
                break
        batch.commit()

    def clear_votes(self) -> None:
        votes = self._client.collection(
            f'games/{self._game_id}/votes').stream()
        with ThreadPoolExecutor(max_workers=self._thread_pool_size) as executor:
            executor.submit(self._delete_vote_fn, votes)

    def find_matchmaker_games(self, region="US") -> list:
        games_list = []
        db = self._client
        games = db.collection('games').where('country_code', '==', region).where(
            'game_has_started', '==', False).stream()

        for game in games:
            try:
                # Avoid checking in query to allow for key to not exist
                if not game.to_dict().get("to_be_deleted"):
                    games_list.append(game)
            except:
                pass
        return games_list

    def ballot(self, player_id: str, challenge_id: str, options: Dict[str, str], is_for_win: bool = False) -> None:
        ballot_ref = self._client.collection(
            f'games/{self._game_id}/players/{player_id}/ballots'
        ).document()
        ballot_ref.set(
            {
                'challenge_id': challenge_id if challenge_id else '',
                'options': options,
                'timestamp': time.time(),
                'is_for_win': is_for_win
            }
        )
        return FirestoreData(ballot_ref.get())

    def find_ballot(self, player: Player) -> Iterable[Ballot]:
        query = self._client.collection(
            f'games/{self._game_id}/players/{player.id}/ballots').order_by(
                'timestamp', direction=Query.DESCENDING)
        return FirestoreBallot(query.get()[0])

    def find_player(self, phone_number: str) -> Optional[Player]:
        query = self._client.collection(
            f'games/{self._game_id}/players'
        ).where('phone_number', '==', phone_number)
        players = query.get()
        if len(players) > 0:
            return FirestorePlayer(players[0])
        else:
            return None
